### PR TITLE
Migrate asr bleu

### PR DESCRIPTION
**Pull request by CalderJohnson**
## Why ?

Why do we need to implement this feature ? What is the use case ?

asr_bleu is a program that is currently a part of fairseq. It evaluates the performance of an ASR (automatic speech recognition) model by using the BLEU (bilingual evaluation understudy) metric. If integrated into the stopes library, it will be able to leverage the many advantages stopes offers such as easy configuration and parallelism (currently it runs entirely in serial with no parallelism). The migration of this pipeline will be a powerful and useful addition to the stopes toolkit and expand the functionality of the library even further. 

## How ?

Document the technical decisions you made.

We set up the asr_bleu pipeline in stopes by taking the original code and moving its configurations into a hydra based setup. We also created a retrieve_data module to encapsulate the code the composes the evaluation data. The goal is to take the existing pipeline and integrate it into stopes in a way that allows for easy configuration and parallelism.

If some parts are WIP, please explicit them here.

The pipeline is in a minimum viable product state. The rest of the pipeline's functionality is still serial, and modules like the transcribe_audio module still need to be put into stopes modules. There will be much more work done to finish the pipeline's functionality and extensively test and optimize it. The pipeline in its entirety should be considered WIP.

## Test plan

We tested the pipeline using this sample data and the french huggingface asr model. Inside the folder is also data for an english test. However, the english asr model is from fairseq, and there is a compatibility issue we are still resolving with the fairseq models.

https://drive.google.com/drive/folders/1vJNKsfQLz5Vfl8yXF_-cdMF-MgxBN_ZZ?usp=sharing

### CODE REVIEW

File: stopes/pipelines/asr_bleu/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+fairseq==0.12.2
```
Comment (Mortimerp9) [inline stopes/pipelines/asr_bleu/requirements.txt:1]: we shouldn't have separate requirements, you can add a sub-group in the main pyproject.toml file if you need to (but probably the `speech` one already covers all that you need)

Comment (CalderJohnson) [reply stopes/pipelines/asr_bleu/requirements.txt]: Good point. Some dependencies such as sacrebleu aren't currently covered under `speech_mining`, so I added a new sub-group with all the dependencies that wouldn't be present in a base installation of stopes. Let me know if I should organize it in a different way.

File: stopes/pipelines/asr_bleu/configs.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+@dataclass
+class CorporaConfig:
+    lang: str
+    audio_dirpath: str
+    reference_path: str
+    reference_format: str
+    reference_tsv_column: str = None
+    audio_format: str = "n_pred.wav"
+    results_dirpath: str = None
+    transcripts_path: str = None
```
Comment (kauterry) [inline stopes/pipelines/asr_bleu/configs.py:12]: Shouldn't some of these attributes be `List[str]` to read in multiple datasets?

File: stopes/pipelines/asr_bleu/retrieve_data.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+                audio_list.append(audio_fp)
+        else:
+            raise NotImplementedError
+
+        return audio_list
+    
+    def _extract_text_for_eval(self, references_filepath: str, reference_format: str, reference_tsv_column: str = None):
+        """Extract sentences for reference"""
+        if reference_format == "txt":
+            reference_sentences = open(references_filepath, "r").readlines()
```
Comment (kauterry) [inline stopes/pipelines/asr_bleu/retrieve_data.py:79]: Please use the `open()` from stopes.core.utils, in general it is good to scour stopes/core/utils.py to see if there are already functions which you can use.

File: stopes/pipelines/asr_bleu/retrieve_data.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    def _extract_audio_for_eval(self, audio_dirpath: str, audio_format: str):
+        """Extract audio file paths for subsequent transcription"""
+        if audio_format == "n_pred.wav":
+            """
+            The assumption here is that 0_pred.wav corresponds to the reference at line position 0 from the reference manifest
+            """
+            audio_list = []
+            audio_fp_list = glob((Path(audio_dirpath) / "*_pred.wav").as_posix())
+            audio_fp_list = sorted(
+                audio_fp_list, key=lambda x: int(os.path.basename(x).split("_")[0])
```
Comment (lpw0) [inline stopes/pipelines/asr_bleu/retrieve_data.py:53]: We will assume that audios and reference_text in the input configuration are pairs. Will sorting by basename here disrupt this matching?

File: stopes/pipelines/asr_bleu/retrieve_data.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+                    assert len(audio_fp) == 1
+                    audio_fp = audio_fp[0]
+
+                audio_list.append(audio_fp)
+        else:
+            raise NotImplementedError
+
+        return audio_list
+    
+    def _extract_text_for_eval(self, references_filepath: str, reference_format: str, reference_tsv_column: str = None):
```
Comment (kauterry) [inline stopes/pipelines/asr_bleu/retrieve_data.py:76]: Can you also type what these functions are returning? I guess `-> tp.List[str]` in this case?

File: stopes/pipelines/asr_bleu/transcribe_audio.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    def array(self):
+        return self.config.transcribe_audio_jobs
+    
+    def requirements(self) -> Requirements:
+        return Requirements(
+            nodes=1,
+            tasks_per_node=1,
+            gpus_per_node=0,
+            cpus_per_task=1,
+            timeout_min=720,
```
Comment (kauterry) [inline stopes/pipelines/asr_bleu/transcribe_audio.py:39]: Maybe change this to 24 * 60, to be safe.

File: stopes/pipelines/asr_bleu/compute_asr_bleu.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        )
+
+        # 4. Compute BLEU score
+        logging.info("Computing BLEU scores...")
+        bleu_scores = []
+        for i, prediction_transcripts in enumerate(transcribed_audio):
+            references = eval_manifests[i]["reference"]
+            bleu_score = sacrebleu.corpus_bleu(prediction_transcripts, [references])
+            bleu_scores.append(bleu_score)
+            print(bleu_score)
```
Comment (kauterry) [inline stopes/pipelines/asr_bleu/compute_asr_bleu.py:50]: Perhaps write this to a file in the `output_dir`.

File: stopes/pipelines/asr_bleu/compute_asr_bleu.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+from stopes.pipelines.asr_bleu.retrieve_data import retrieve_data
+
+import hydra
+import sacrebleu
+
+class AsrBleu:
+    def __init__(self, config: AsrBleuConfig):
+        self.config = config
+        self.launcher = hydra.utils.instantiate(self.config.launcher)
+        self.logger = logging.getLogger("asr_bleu")
```
Comment (kauterry) [inline stopes/pipelines/asr_bleu/compute_asr_bleu.py:16]: `import stopes.core.utils`, and `utils.ensure_dir(self.config.output_dir)` here.

File: stopes/pipelines/asr_bleu/transcribe_audio.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+        asr_model = ASRGenerator(iteration_value.asr_config)
+
+        prediction_transcripts = []
+    
+        #TODO: reimplement tqdm progress bar with the dictionary instead of the dataframe
+        #for _,  eval_pair in tqdm(
+        #    iteration_value.eval_manifest.iterrows(),
+        #    desc="Transcribing predictions",
+        #    total=len(iteration_value.eval_manifest["prediction"]),
+        #):
```
Comment (kauterry) [inline stopes/pipelines/asr_bleu/transcribe_audio.py:176]: Make sure you exclude the commented out lines in your PR.

File: stopes/pipelines/asr_bleu/utils.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+        self.model = model[0]
+        self.tokens = tokens
+
+        if "|" in tokens:
+            self.sil_token = "|"
+        else:
+            self.sil_token = tokens[
+                2
+            ]  # use eos as silence token if | not presented e.g., Hok ASR model
+        print(f"Inferring silence token from the dict: {self.sil_token}")
```
Comment (kauterry) [inline stopes/pipelines/asr_bleu/utils.py:206]: Use `logger.info()` as opposed to `print`, everywhere.

File: stopes/pipelines/asr_bleu/compute_asr_bleu.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+    async def run(self):
+        # 1. Retrieve ASR configuration 
+        logging.info("Setting up ASR model...")
```
Comment (kauterry) [inline stopes/pipelines/asr_bleu/compute_asr_bleu.py:20]: Should be `logger.info()` right?

File: stopes/pipelines/asr_bleu/compute_asr_bleu.py (hunk 8 above)
Comment (kauterry) [inline stopes/pipelines/asr_bleu/compute_asr_bleu.py:16]: You can make `logger` a global variable in this file for convenience.

**Review by kauterry — CHANGES_REQUESTED** [review]
Excellent job putting this together! I guess you've implemented the pipeline for a single dataset, now you can work on making this work for multiple datasets in parallel. Should be the next step from here, on the Stopes side of things. I'll leave @lpw0 to review the more ASR-specific parts of the PR.

**Review by lpw0 — CHANGES_REQUESTED** [review]

File: stopes/pipelines/asr_bleu/compute_asr_bleu.py (hunk 11 above)
Comment (CalderJohnson) [reply stopes/pipelines/asr_bleu/compute_asr_bleu.py]: Thanks, resolved with https://github.com/facebookresearch/stopes/pull/52/commits/a2682411272445a780ca428074c2f3c42ecd8b2e

File: stopes/pipelines/asr_bleu/compute_asr_bleu.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        asr_config = retrieve_asr_config(self.config.corpora.lang, self.config.asr_version, json_path="/home/calderj/Documents/Coding/MLH/stopes/stopes/pipelines/asr_bleu/conf/asr_model/asr_model_cfgs.json")
```
[the remark this answers is not in the document]
Comment (CalderJohnson) [reply stopes/pipelines/asr_bleu/compute_asr_bleu.py]: Thanks, resolved with https://github.com/facebookresearch/stopes/pull/52/commits/0f5c137422fdb3076de4c5e6870dff90389ff4ce

[the remark this answers is not in the document]
Comment (CalderJohnson) [reply stopes/pipelines/asr_bleu/compute_asr_bleu.py]: Thanks, resolved with https://github.com/facebookresearch/stopes/pull/52/commits/0f5c137422fdb3076de4c5e6870dff90389ff4ce

File: stopes/pipelines/asr_bleu/transcribe_audio.py (hunk 9 above)
Comment (CalderJohnson) [reply stopes/pipelines/asr_bleu/transcribe_audio.py]: Thanks, resolved in https://github.com/facebookresearch/stopes/pull/52/commits/c6e830f1225545544f9d104b27874c5479925491

File: stopes/pipelines/asr_bleu/transcribe_audio.py [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+
+from stopes.core.launcher import Launcher
+from stopes.core.stopes_module import Requirements, StopesModule
+from stopes.pipelines.asr_bleu.utils import ASRGenerator
+
+@dataclass
+class TranscribeAudioJob:
+    eval_manifest: tp.Dict[str, tp.List] = MISSING
+    asr_config: tp.Dict = MISSING
+
```
[the remark this answers is not in the document]
Comment (CalderJohnson) [reply stopes/pipelines/asr_bleu/transcribe_audio.py:21]: Thanks, resolved in https://github.com/facebookresearch/stopes/pull/52/commits/c6e830f1225545544f9d104b27874c5479925491 and https://github.com/facebookresearch/stopes/pull/52/commits/d720e65003c328f73eb256b9354623591766d1e3

File: stopes/pipelines/asr_bleu/retrieve_data.py (hunk 4 above)
Comment (CalderJohnson) [reply stopes/pipelines/asr_bleu/retrieve_data.py:54]: Could you explain more about your concern? My understanding is that basename is just the isolated file name (without the rest of the path). The audio files are named `n_pred.wav`, then they're sorted by their number `n`. As long as there are the same number of reference lines as audio files, and the names are formatted properly, how would the matching get disrupted?

Comment (lpw0) [reply stopes/pipelines/asr_bleu/retrieve_data.py:54]: then, nvm, it looks like the matching will be correct as long as the input reference rows with the same order as the wav name. Actually, the generation part (before the ASR-BLEU) would generate the *_pred.wav according to the reference column and the order would be helpful for debugging.

File: pyproject.toml [hunk 14]
```diff
@@ -88,6 +88,13 @@ classifiers=[
     "torch",
     # fairseq is also a dependency, but it must be installed separately
   ]
+  asr_bleu = [
+    "pandas==1.4.3",
```
Comment (avidale) [inline pyproject.toml:92]: Do we really need such precise versions?

File: stopes/pipelines/asr_bleu/utils.py [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+def retrieve_asr_config(lang_key: str, asr_version: str, json_path: str) -> dict:
+    if len(lang_key) !=3:
+        raise ValueError(f"'{lang_key}' lang key for language type must be 3 characters!")
+
+    with open(json_path, "r") as f:
+        asr_model_cfgs = json.load(f)
+    return asr_model_cfgs[lang_key][asr_version]
+
+
+class ASRGenerator(object):
```
Comment (avidale) [inline stopes/pipelines/asr_bleu/utils.py:43]: If the class represents a generator, shouldn't it contain the actual method for generation (i.e. transcription)?

File: stopes/pipelines/asr_bleu/transcribe_audio.py [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+        emissions = emissions.cpu()
+        results = decoder(emissions)
+
+        # assuming the lexicon-free decoder and working with tokens
+        hypo = decoder.idxs_to_tokens(results[0][0].tokens)
+        hypo = post_process_fn(hypo)
+
+        return hypo
+
+    def _merge_tailo_init_final(self, text):
```
Comment (avidale) [inline stopes/pipelines/asr_bleu/transcribe_audio.py:114]: This method doesn't seem to reuse any properties or methods of the `TranscribeAudio`, so probably it should be turned into a standalone function. 

Also, it's totally unclear what it actually does. Maybe, a few examples of inputs and outputs could illustrate it?

File: stopes/pipelines/asr_bleu/transcribe_audio.py [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+                continue
+            last_syllable += sp
+            if sp[-1].isnumeric():
+                results.append(last_syllable)
+                last_syllable = ""
+        if last_syllable != "":
+            results.append(last_syllable)
+        return " ".join(results)
+
+    def _remove_tone(self, text):
```
Comment (avidale) [inline stopes/pipelines/asr_bleu/transcribe_audio.py:132]: Is this used anywhere?

File: stopes/pipelines/asr_bleu/compute_asr_bleu.py [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+import hydra
+import sacrebleu
+
+logger = logging.getLogger("stopes.asr_bleu")
+
+class AsrBleu:
+    def __init__(self, config: AsrBleuConfig):
+        self.config = config
+        self.launcher = hydra.utils.instantiate(self.config.launcher)
+        utils.ensure_dir(self.config.output_dir)
```
Comment (avidale) [inline stopes/pipelines/asr_bleu/compute_asr_bleu.py:19]: If I understand correctly, this script will print the BLEU score, but the transcriptions won't be saved anywhere. 
Does it mean that `output_dir` is going to be unused?

File: stopes/pipelines/asr_bleu/transcribe_audio.py (hunk 17 above)
Comment (CalderJohnson) [reply stopes/pipelines/asr_bleu/transcribe_audio.py:132]: No, it was left over from the old codebase where it was also not used. I'll remove it for now.

Comment (CalderJohnson) [reply stopes/pipelines/asr_bleu/transcribe_audio.py]: Resolved with https://github.com/facebookresearch/stopes/pull/52/commits/c422fc7985fe8c84f42b7829d3c2efaadf3eef9a

File: pyproject.toml (hunk 14 above)
Comment (CalderJohnson) [reply pyproject.toml]: You're right, big libraries like pandas will be backwards compatible. I'll just keep the precise version for sacrebleu, as we've had issues before with libraries within the FAIR ecosystem needing to be specific versions. Resolved with https://github.com/facebookresearch/stopes/pull/52/commits/c422fc7985fe8c84f42b7829d3c2efaadf3eef9a

File: stopes/pipelines/asr_bleu/utils.py (hunk 15 above)
Comment (CalderJohnson) [reply stopes/pipelines/asr_bleu/utils.py]: Good point. We moved the generation code to transcribe_audio, but the name remained unchanged. Would ASRContainer be a more suitable name? https://github.com/facebookresearch/stopes/pull/52/commits/5e9187eb43e0a29d044ac4e49beb8a41ccbce3a8

File: stopes/pipelines/asr_bleu/utils.py [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+        """
+        Update the download progress
+        """
+        if tsize is not None:
+            self.total = tsize
+        self.update(b * bsize - self.n)
+
+
+def retrieve_asr_config(lang_key: str, asr_version: str, json_path: str) -> dict:
+    if len(lang_key) !=3:
```
[the remark this answers is not in the document]
Comment (CalderJohnson) [reply stopes/pipelines/asr_bleu/utils.py]: Ran a static type checker and a linter. Resolved all errors that weren't just it not seeing my imports. https://github.com/facebookresearch/stopes/pull/52/commits/f9c6242477bcfa93bcd335c87c5d4c6b10fef2ef

File: stopes/pipelines/asr_bleu/retrieve_data.py [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+        eval_manifest = {
+            "prediction": audio_list,
+            "reference": reference_sentences,
+        }
+
+        return eval_manifest
+
+
+async def retrieve_data(
+    datasets: tp.List[tp.Tuple[str, str, str, str, str]],
```
[the remark this answers is not in the document]
Comment (CalderJohnson) [reply stopes/pipelines/asr_bleu/retrieve_data.py]: Good point. My updated setup allowing for parallel datasets does it this way. https://github.com/facebookresearch/stopes/pull/52/commits/abc9d0055266365a22eaadc76bd4f4fcad73c8ec

File: .gitignore [hunk 21]
```diff
@@ -136,3 +136,6 @@ demo/*.gz
 demo/models/
 demo/mining/*.gz
 demo/mining/models/
+
+# Testing configs
```
[the remark this answers is not in the document]
Comment (CalderJohnson) [reply .gitignore]: Makes sense. Fixed https://github.com/facebookresearch/stopes/pull/52/commits/42827f1eaf0d4f7b9a5f60a9bdcacaaf86305b04

File: stopes/pipelines/asr_bleu/transcribe_audio.py (hunk 16 above)
Comment (lpw0) [reply stopes/pipelines/asr_bleu/transcribe_audio.py:143]: This function is for post processing the ASR output to merge the phoneme initial and finals back to syllable units and it is specifically for Hokkien, details in https://github.com/fairinternal/ust_common/pull/47

**Pull request closed without merging**